### PR TITLE
CLOUDSTACK-9827: Storage tags stored in multiple places

### PR DESCRIPTION
**Pull request by nvazquez**
Issue description: https://issues.apache.org/jira/browse/CLOUDSTACK-9827

### Fixes
- Create Primary Storage: Persist tags into `storage_pool_tags` instead of `storage_pool_details`
- List Storage Tags: Queries `storage_pool_tags` table instead of `storage_tag_view`
- Find Storage Pools by Tags using proper DAO
- Remove storage tags after deleting Primary Storage
- Remove unused `StorageTagDao`, `StorageTagDaoImpl`, `StorageTagVO` and `storage_tag_view`

### CODE REVIEW

**Comment by serg38** [comment]
@mike-tutkowski @karuturi @koushik-das @rafaelweingartner  Please review the fix for 4.10 blocker

**Comment by serg38** [comment]
@nvazquez Should we simply replace storage_tag_view instead of switching queries over to another view?

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rafaelweingartner** [comment]
@nvazquez I am assuming the change needed to fix the bug is line 135 at PrimaryDataStoreHelper.java?
The other changes are regarding code conformity

**Comment by nvazquez** [comment]
Hi @rafaelweingartner, you're right, it was basically that fix. 
I've pushed another commit due to issue reported by @mike-tutkowski in mailing list:

I have an NFS SR as primary storage under CloudStack with a storage tag of NFS-A. I have a compute offering with a matching storage tag. I can’t seem to get a VM to spin up, however: It says insufficient capacity. The CPU, MHz, and memory are all low (and what I typically use), so I think the problem is with matching the storage tag.

**Comment by serg38** [comment]
@nvazquez Is storage_tag_view still in use after this change that reference old way of retrieving the tags? If not should we remove it?  E.g. host tags don't have a separate view.

**Comment by nvazquez** [comment]
@serg38 actually is not being used anymore, I'll add removalof the view on last commit

**Comment by nvazquez** [comment]
Added these items to description:
- Find Storage Pools by Tags using proper DAO
- Remove unused `StorageTagDao`, `StorageTagDaoImpl` and `storage_tag_view`

**Comment by serg38** [comment]
@rhtyd @borisstoyanov Can you kick off new tests for this release blocker?

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by serg38** [comment]
@mike-tutkowski Can you test this fix in your environment?

**Comment by mike-tutkowski** [comment]
I should be able to do so over the weekend.

On Mar 10, 2017, at 11:26 AM, serg38 <notifications@github.com<mailto:notifications@github.com>> wrote:


@mike-tutkowski<https://github.com/mike-tutkowski> Can you test this fix in your environment?

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java [hunk 1]
```diff
@@ -77,4 +90,71 @@ public void deleteTags(long poolId) {
         txn.commit();
     }
 
+    @Override
+    public List<StoragePoolTagVO> searchByIds(Long... stIds) {
+        String batchCfg = _configDao.getValue("detail.batch.query.size");
```
Comment (rafaelweingartner) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: @nvazquez what about extracting lines 95-97 to a method?
Then, we can have test cases and some doc for it. Especially for that 2000 magic number there.

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java [hunk 2]
```diff
@@ -77,4 +90,71 @@ public void deleteTags(long poolId) {
...
+
+        if (stIds.length > detailsBatchSize) {
+            while ((curr_index + detailsBatchSize) <= stIds.length) {
+                Long[] ids = new Long[detailsBatchSize];
+
+                for (int k = 0, j = curr_index; j < curr_index + detailsBatchSize; j++, k++) {
+                    ids[k] = stIds[j];
+                }
+
+                SearchCriteria<StoragePoolTagVO> sc = StoragePoolIdsSearch.create();
```
Comment (rafaelweingartner) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: @nvazquez lines 105-119 are the same as 128-142.
What about extracting them for a method?

File: engine/schema/src/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImpl.java [hunk 3]
```diff
@@ -409,15 +460,13 @@ public StoragePoolVO persist(StoragePoolVO pool, Map<String, String> details) {
             sc.and(sc.entity().getScope(), Op.EQ, ScopeType.ZONE);
             return sc.list();
         } else {
-            Map<String, String> details = tagsToDetails(tags);
-
-            StringBuilder sql = new StringBuilder(ZoneWideDetailsSqlPrefix);
+            StringBuilder sql = new StringBuilder(ZoneWideTagsSqlPrefix);
```
Comment (rafaelweingartner) [inline engine/schema/src/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImpl.java]: @nvazquez what about extracting lines 463-469 to a method? Then, we can create doc and most of all test cases. This is a pretty tricky bit of code (Especially line 468).

**Comment by mike-tutkowski** [comment]
I have run the following use cases successfully:

PS = Primary Storage
ST = Storage Tag
CO = Compute Offering
DO = Disk Offering
V = Volume

Create PS_1 with ST NFS-A

Create CO_1 with ST NFS-A
Create CO_2 with ST NFS-B

Create VM_1 with CO_1 (VM starts)
Create VM_2 with CO_2 (VM fails to find storage)

Edit ST of PS_1 from NFS-A to NFS-B

Create VM_2 with CO_1 (VM fails to find storage)
Create VM_2 with CO_2 (VM starts)

Create DO_1 with ST NFS-A
Create DO_2 with ST NFS-B

Create V_1 with DO_1 and attach to VM_1 (fails to find storage)
Create V_2 with DO_2 and attach to VM_1

Edit ST of PS_1 from NFS-B to NFS-A

Create V_1 with DO_1 and attach to VM_1

Create PS_2 with no ST

Edit ST of PS_2 to NFS-B

Create V_3 with DO_1 and attach to VM_2 (lands on expected PS)
Create V_4 with DO_2 and attach to VM_2 (lands on expected PS)

Create PS_3 with ST NFS-B

Migrate V_4 from PS_2 to PS_3 (PS_1 is not an option)
Migrate V_4 from PS_3 to PS_2 (PS_1 is not an option)

Migrate V_3 from PS_1 to PS_2 (Neither PS_2 nor PS_3 is an option)

Migrate root V of VM_1 from PS_1 to PS_2 (Neither PS_2 nor PS_3 is an option)

Migrate root V of VM_2 from PS_1 to PS_3 (PS_2 and PS_3 are options)

listStorageTags API returned expected results (I ran this at several points along the way)

**Comment by karuturi** [comment]
@nvazquez I think any of the existing marvin tests didnt catch this bug. Is it possible to add a marvin test for it?

**Comment by nvazquez** [comment]
@mike-tutkowski awesome, thanks for testing this PR!

@rafaelweingartner thanks for reviewing, I'll work on changes proposed

@karuturi sure, I'll work on it, thanks!

**Comment by nvazquez** [comment]
@rafaelweingartner I pushed changes and squashed my commits as it could be easier to review. I also added unit tests for new methods

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java (hunk 1 above)
Comment (nvazquez) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: Done, thanks!

Comment (nvazquez) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: About number 2000, I assumed it was a default value for that configuration

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java (hunk 2 above)
Comment (nvazquez) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: Done, created method `searchForStoragePoolIdsInternal`

File: engine/schema/src/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImpl.java (hunk 3 above)
Comment (nvazquez) [inline engine/schema/src/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImpl.java]: Thanks for pointing this out, I had missed it out. Created methods `getSqlPreparedStatement` and `searchStoragePoolsPreparedStatement` which are called from many methods and allow storage pool retrieval

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java [hunk 4]
```diff
@@ -77,4 +90,68 @@ public void deleteTags(long poolId) {
...
+    }
+
+    /**
+     * Search storage pools
+     * @param currIndex current index
+     * @param batchSize batch size
+     * @param stIds storage tags array
+     * @param uvList
+     */
+    protected void searchForStoragePoolIdsInternal(int currIndex, int batchSize, Long[] stIds, List<StoragePoolTagVO> uvList) {
```
Comment (rafaelweingartner) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: What about improving this java doc here?
Something like:

>  Search for storage pools based on their IDs. The search is executed in batch, this means that we will load a batch of size #getDetailsBatchSize  @link{StoragePoolTagVO} at each time. The loaded storage pools are added in the uvList parameter.

The links to classes, methods and parameters in the Java doc would have to be fixed (I wrote it without using the IDE and I did not know the proper markup)

What do you think about the "@param" notation? I normally prefer not to use them if the onyl thing I will write is the parameter name such as `@param uvList`

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java [hunk 5]
```diff
@@ -77,4 +90,68 @@ public void deleteTags(long poolId) {
...
+        }
+    }
+
+    /**
+     * Retrieve {@code detail.batch.query.size} configuration value. If not available, return the default value (2000)
+     * @return detail.batch.query.size configuration value
+     */
+    protected int getDetailsBatchSize() {
+        String batchCfg = _configDao.getValue("detail.batch.query.size");
+        return batchCfg != null ? Integer.parseInt(batchCfg) : 2000;
```
Comment (rafaelweingartner) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: What about extracting the `2000` to a constant in the class?

File: engine/schema/test/com/cloud/storage/dao/StoragePoolTagsDaoImplTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    ConfigurationDao _configDao;
+    @Mock
+    SearchBuilder<StoragePoolTagVO> StoragePoolIdsSearch;
+
+    @Spy
+    @InjectMocks
+    private StoragePoolTagsDaoImpl _storagePoolTagsDaoImpl = new StoragePoolTagsDaoImpl();
+
+    @Mock
+    static StoragePoolTagVO storagePoolTag1;
```
Comment (rafaelweingartner) [inline engine/schema/test/com/cloud/storage/dao/StoragePoolTagsDaoImplTest.java]: I am curious here, is there any reason to declare the variables here `static`?

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java (hunk 4 above)
Comment (nvazquez) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: Excellent improvement, I copied almost exactly as you wrote it :) I also added some description on `uvList` parameter, which also renamed to `pools`

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java (hunk 5 above)
Comment (nvazquez) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: Done, thanks!

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java [hunk 7]
```diff
@@ -77,4 +92,71 @@ public void deleteTags(long poolId) {
...
 
+    @Override
+    public List<StoragePoolTagVO> searchByIds(Long... stIds) {
+        final int detailsBatchSize = getDetailsBatchSize();
+
+        // query details by batches
+        List<StoragePoolTagVO> uvList = new ArrayList<StoragePoolTagVO>();
+        int curr_index = 0;
+
+        if (stIds.length > detailsBatchSize) {
```
Comment (rafaelweingartner) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: Correct me if I am wrong, but this `if` is not needed.
Let's assume the configuration:
```
batchSize=2000
current_index=0
lengthOfStIds=100
```

`curr_index + detailsBatchSize = 0 + 2000`, which is not less than the size of the array `(100)`. Therefore, the while is not executed. Then, the pools will be loaded at lines 100-112

File: engine/schema/test/com/cloud/storage/dao/StoragePoolTagsDaoImplTest.java (hunk 6 above)
Comment (nvazquez) [inline engine/schema/test/com/cloud/storage/dao/StoragePoolTagsDaoImplTest.java]: Not really, thinking about it there's no why to make them static, I removed static declaration from them

File: engine/schema/test/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImplTest.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    StoragePoolHostDao _hostDao;
+    @Mock
+    StoragePoolTagsDao _tagsDao;
+
+    @Spy
+    @InjectMocks
+    private static PrimaryDataStoreDaoImpl primaryDataStoreDao = new PrimaryDataStoreDaoImpl();
+
+    @Mock
+    static StoragePoolVO storagePoolVO;
```
Comment (rafaelweingartner) [inline engine/schema/test/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImplTest.java]: Here you have other static variable in a test case that does not need to be static.

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java (hunk 7 above)
Comment (nvazquez) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: I think that `if` is needed to control the max query size. I agree with the example you provided, but let's say for example that lengthOfStIds is greater that batchSize. If we remove that `if`, we will load pools on lines 100-112 but with a query size greater that batchSize (defined in line 111)

File: engine/schema/test/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImplTest.java (hunk 8 above)
Comment (nvazquez) [inline engine/schema/test/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImplTest.java]: Thanks! I removed it

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java (hunk 7 above)
Comment (nvazquez) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: Hi @rafaelweingartner, I assumed that you meant deleting the whole `if` block (lines 103-108), is it correct or you mean just deleting line 103 (and 108)?

Comment (rafaelweingartner) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: Ah, no this is not what I meant.
I meant only removing the `if` and letting the `while` directly there.
Just deleting line 103

Comment (nvazquez) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: Nice, I agree with you that deleting the line can be removed. Sorry for the confusion, I'll try adding some unit tests for this change. Thanks!

Comment (rafaelweingartner) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: No problem ;)

It is awesome the changes you have done here. The code became cleaner, with good documentation and on top of that, we now have unit tests to guarantee that the methods are workings as expected.

Great job @nvazquez, well done!

PS: Sorry if I have been too picky with the code changes.

**Comment by karuturi** [comment]
@nvazquez any update?

**Comment by nvazquez** [comment]
Hi @karuturi, I've been working on marvin tests, I hope posting them today

File: engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java (hunk 7 above)
Comment (nvazquez) [inline engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java]: Done! Thanks a lot @rafaelweingartner!

**Comment by nvazquez** [comment]
@karuturi I added marvin tests to simulate tests performed by @mike-tutkowski.

This are results in our env:
````
[root@ussarlabcsmgt41 cloudstack]# cat /tmp//MarvinLogs//011CTF/results.txt
Test primary storage pools - XEN. Not Supported for kvm,hyperv,vmware ... SKIP: iscsi primary storage not supported on kvm, VMWare, Hyper-V, or LXC
Test primary storage pools - XEN, KVM, VMWare. Not Supported for hyperv ... === TestName: test_01_primary_storage_nfs | Status : SUCCESS ===
ok
Test Deploy VMS using different Service Offerings with Storage Tags ... === TestName: test_01_deploy_vms_storage_tags | Status : SUCCESS ===
ok
Test edit Storage Tags ... === TestName: test_02_edit_primary_storage_tags | Status : SUCCESS ===
ok
Test Volume migration options for Storage Pools with different Storage Tags ... SKIP: Skipping test as it is not running on simulator

----------------------------------------------------------------------
Ran 5 tests in 295.996s

OK (SKIP=2)
````

@rafaelweingartner @mike-tutkowski @karuturi can you please review marvin tests added?

**Comment by karuturi** [comment]
Thanks @nvazquez Waiting for LGTMs

**Comment by serg38** [comment]
@nvazquez Looks like Travis cant SSH into management server during the test .

2017-03-17 19:26:58,672 - DEBUG - mount -t nfs nfs:/export/automation/1/testprimary /mnt/marvin-1b4190cb-b9f5-4441-9bdc-bd89e9b7567b
2017-03-17 19:27:13,468 - CRITICAL - EXCEPTION: test_03_migration_options_storage_tags: ['Traceback (most recent call last):\n', '  File "/opt/python/2.7.12/lib/python2.7/unittest/case.py", line 329, in run\n    testMethod()\n', '  File "/home/travis/.local/lib/python2.7/site-packages/marvin/lib/decoratorGenerators.py", line 30, in test_wrapper\n    return test(self, *args, **kwargs)\n', '  File "/home/travis/build/apache/cloudstack/test/integration/smoke/test_primary_storage.py", line 559, in test_03_migration_options_storage_tags\n    log_lvl=logging.INFO\n', '  File "/home/travis/.local/lib/python2.7/site-packages/marvin/sshClient.py", line 81, in __init__\n    raise internalError("SSH Connection Failed")\n', 'internalError: SSH Connection Failed\n']

**Comment by nvazquez** [comment]
Thanks @serg38, we are using `mgtSvr` details provided in .cfg file, should we use these for Marvin too?

**Comment by karuturi** [comment]
@mike-tutkowski can you do a final review please?

**Comment by mike-tutkowski** [comment]
I should have a chance to look through this later tonight. In the meanwhile, perhaps we can re-push the code to kick off Travis again (since its most recent run has a failure). Thanks!

**Comment by mike-tutkowski** [comment]
OK, this LGTM (as long as Travis shows green before we merge). Thanks!

**Comment by nvazquez** [comment]
Thanks @mike-tutkowski! I pushed force to kick off Travis again

**Comment by nvazquez** [comment]
@karuturi I refactored last marvin test which was failing on Travis. These are results in our env:

````
[root@ussarlabcsmgt41 cloudstack]# cat /tmp//MarvinLogs//4GSNSY/results.txt
Test primary storage pools - XEN. Not Supported for kvm,hyperv,vmware ... SKIP: iscsi primary storage not supported on kvm, VMWare, Hyper-V, or LXC
Test primary storage pools - XEN, KVM, VMWare. Not Supported for hyperv ... === TestName: test_01_primary_storage_nfs | Status : SUCCESS ===
ok
Test Deploy VMS using different Service Offerings with Storage Tags ... === TestName: test_01_deploy_vms_storage_tags | Status : SUCCESS ===
ok
Test Edit Storage Tags ... === TestName: test_02_edit_primary_storage_tags | Status : SUCCESS ===
ok
Test Volume migration options for Storage Pools with different Storage Tags ... === TestName: test_03_migration_options_storage_tags | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 5 tests in 442.216s

OK (SKIP=1)
````

@rhtyd @borisstoyanov @jburwell on `test_primary_storage.py` we are creating 2 primary storages, using urls in `"nfs"` and `"nfs2"` test data templates. As PR #1961 uses `"nfs2"` template on `test_snapshots.py`, should it be a problem for BlueOrangutan?

**Comment by nvazquez** [comment]
@karuturi Travis is now failing as it doesn't find key "nfs2"
````
2017-03-24 17:33:45,621 - CRITICAL - EXCEPTION: test_03_migration_options_storage_tags: ['Traceback (most recent call last):\n', '  File "/opt/python/2.7.12/lib/python2.7/unittest/case.py", line 329, in run\n    testMethod()\n', '  File "/home/travis/.local/lib/python2.7/site-packages/marvin/lib/decoratorGenerators.py", line 30, in test_wrapper\n    return test(self, *args, **kwargs)\n', '  File "/home/travis/build/apache/cloudstack/test/integration/smoke/test_primary_storage.py", line 520, in test_03_migration_options_storage_tags\n    self.services["nfs2"],\n', "KeyError: 'nfs2'\n"]
````
It is introduced in PR #1961, can these two PRs be merged together?

**Comment by karuturi** [comment]
Thank you all. I will merge #1961 and #1994 now

File: engine/schema/test/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImplTest.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    @Test
+    public void testFindPoolsByDetailsOrTagsInternalStorageTagsType() {
+        List<StoragePoolVO> storagePools = primaryDataStoreDao.findPoolsByDetailsOrTagsInternal(DATACENTER_ID, POD_ID, CLUSTER_ID, SCOPE, SQL_VALUES, ValueType.TAGS, STORAGE_TAGS_ARRAY.length);
+        assertEquals(Arrays.asList(storagePoolVO), storagePools);
+        verify(primaryDataStoreDao).getSqlPreparedStatement(
+                primaryDataStoreDao.TagsSqlPrefix, primaryDataStoreDao.TagsSqlSuffix, SQL_VALUES, CLUSTER_ID);
+        String expectedSql = primaryDataStoreDao.TagsSqlPrefix + SQL_VALUES + primaryDataStoreDao.TagsSqlSuffix;
+        verify(primaryDataStoreDao).searchStoragePoolsPreparedStatement(expectedSql, DATACENTER_ID, POD_ID, CLUSTER_ID, SCOPE, STORAGE_TAGS_ARRAY.length);
```
Comment (anshul1886) [inline engine/schema/test/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImplTest.java]: @nvazquez With these types of test we are kind of forcing the SQL statement to use in implementation. Do you think this is correct thing to test? What if somebody wants to rewrite in different way altogether to achieve the same result?

Comment (nvazquez) [inline engine/schema/test/org/apache/cloudstack/storage/datastore/db/PrimaryDataStoreDaoImplTest.java]: Hi @anshul1886,
I understand your point of view and I think it is valid. However, from my point of view these unit tests are covering internal methods used in the implementation. For example, for method `String getSqlValuesFromStorageTags(String[] tags)` tests provided cover cases when parameter is null, empty or not null. As this method returns a string, it is the goal of unit tests to check if the string returned by the method is the one we expect for each of those cases, and in my opinion, it is being tested properly. I mention this particular method as string returned is a SQL statement. 
The last two tests in this file are testing `findPoolsByDetailsOrTagsInternal`, comparing returned list with the one expected and also verifying that internal methods are invoked with expected parameters, but these parametes are strings (SQL statements). From my point of view, we are not forcing SQL statement to use, we are checking that internal methods are being invoked properly.
In this way, if method is refactored but achieves the same result, we have to modify/remove lines which verifies if internal methods are invoked for the ones that belong to new implementation.
Please let me know if I didn't explain myself, what do you think? I don't mean to be stubborn with this, it was my point of view, and of course I might be wrong. What do you think it has to be refactored in these tests?

**Comment by anshul1886** [comment]
@nvazquez I have this PR #1835 which is making changes in that SQL. Now the the tests written by you are not allowing those changes. How should we go about it?

**Comment by nvazquez** [comment]
Hi @anshul1886, you can modify `verify` lines on tests `testFindPoolsByDetailsOrTagsInternalStorageTagsType` and `testFindPoolsByDetailsOrTagsInternalDetailsType` for the ones that methods will expect, or if you want you can delete them to not be tied to SQL implementation

**Comment by anshul1886** [comment]
@nvazquez Is it ok for you to remove those tests?

**Comment by nvazquez** [comment]
@anshul1886 I would prefer that you delete `verify` lines on those tests, but if you want to remove tests it is ok for me.

**Pull request merged**